web3/contract.py

```diff
@@ -1416,7 +1416,7 @@ def parse_block_identifier(web3, block_identifier):
     elif block_identifier in ['latest', 'earliest', 'pending']:
         return block_identifier
     elif isinstance(block_identifier, bytes) or is_hex_encoded_block_hash(block_identifier):
-        return web3.eth.getBlock(block_identifier).number
```

Comment:
cc @dylanjw @carver @voith just in case you don't see this.  This is a good example of why we should be as diligent as possible to not assume the `AttributeDict` API is usable within our own web3 apis.  Forgive me for preaching to the choir if you're already aware.

Reply:
FWIW, replacing *all* the middlewares as the user did in #1039 is almost definitely not what they wanted to do. There will be a lot of things that work incorrectly, like: fields won't be converted to and from python native types. I'm not suggesting that we should *encourage* attribute syntax, but this is only one of many things that will work in unexpected ways if the user drops all the middleware.